File: scripts/dump_records.py
"""Dump the estimator history CSV as a JSON array of record objects.

Reuses build_csv's reader/HEADERS. Each record: {"row": <1-based>, <field>: value, ...}.
Usage: dump_records.py <history.csv>
"""

import json
import os
import sys

sys.path.insert(0, os.path.dirname(__file__))  # build_csv is in the same dir
import build_csv  # noqa: E402

FIELD_KEYS = build_csv.FIELD_KEYS


def dump(rows):
    out = []
    for i, row in enumerate(rows, start=1):
        if not row:
            continue
        rec = {"row": i}
        for j, fk in enumerate(FIELD_KEYS):
            rec[fk] = row[j] if j < len(row) else ""
        out.append(rec)
    return out


def main(argv):
    if len(argv) != 2:
        sys.exit("Usage: dump_records.py <history.csv>")
    rows = build_csv.read_existing(argv[1])
    sys.stdout.buffer.write(
        (json.dumps(dump(rows), ensure_ascii=False, separators=(",", ":")) + "\n").encode("utf-8")
    )


if __name__ == "__main__":
    main(sys.argv)

File: scripts/dump_records.mjs
// Dump the estimator history CSV as a JSON array of record objects.
// Reuses build_csv's reader/HEADERS. Each record: {"row":N, <field>:value, ...}.
// Usage: dump_records.mjs <history.csv>
import { HEADERS, FIELD_KEYS as _FIELD_KEYS, readExisting } from "./build_csv.mjs";
import { pathToFileURL } from "node:url";

export const FIELD_KEYS = _FIELD_KEYS;

export function dump(rows) {
  const out = [];
  rows.forEach((row, idx) => {
    if (!row || !row.length) return;
    const rec = { row: idx + 1 };
    FIELD_KEYS.forEach((fk, j) => { rec[fk] = row[j] ?? ""; });
    out.push(rec);
  });
  return out;
}

export function main(argv) {
  if (argv.length !== 1) {
    console.error("Usage: dump_records.mjs <history.csv>");
    process.exit(1);
  }
  console.log(JSON.stringify(dump(readExisting(argv[0]))));
}

if (process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href) {
  main(process.argv.slice(2));
}
